projects/gloo/pkg/plugins/local_ratelimit/plugin.go

```diff
@@ -0,0 +1,251 @@
+
+func (p *plugin) NetworkFiltersTCP(params plugins.Params, listener *v1.TcpListener) ([]plugins.StagedNetworkFilter, error) {
+	return generateNetworkFilter(listener.GetOptions().GetLocalRatelimit())
+}
+
+func GenerateHTTPFilter(settings *local_ratelimit.Settings, localRatelimit *local_ratelimit.TokenBucket, stage uint32) (*envoy_extensions_filters_http_local_ratelimit_v3.LocalRateLimit, error) {
+	tokenBucket, err := toEnvoyTokenBucket(localRatelimit)
+	if err != nil {
+		return nil, err
+	}
+	filter := &envoy_extensions_filters_http_local_ratelimit_v3.LocalRateLimit{
+		StatPrefix:  HTTPFilterStatPrefix,
+		TokenBucket: tokenBucket,
+		Stage:       stage,
+	}
+
+	// Do NOT set filter enabled and enforced if the token bucket is not found. This causes it to rate limit all requests to zero
+	if tokenBucket != nil {
+		filter.FilterEnabled = &corev3.RuntimeFractionalPercent{
+			DefaultValue: &envoy_type_v3.FractionalPercent{
+				Numerator:   100,
+				Denominator: envoy_type_v3.FractionalPercent_HUNDRED,
+			},
+		}
+		filter.FilterEnforced = &corev3.RuntimeFractionalPercent{
+			DefaultValue: &envoy_type_v3.FractionalPercent{
+				Numerator:   100,
+				Denominator: envoy_type_v3.FractionalPercent_HUNDRED,
+			},
+		}
+	}
+	// This needs to be set on every virtual service or route that has custom local RL as they default to false and override the HCM level config
+	filter.LocalRateLimitPerDownstreamConnection = settings.GetLocalRateLimitPerDownstreamConnection()
+	if settings.GetEnableXRatelimitHeaders() {
+		filter.EnableXRatelimitHeaders = envoyratelimit.XRateLimitHeadersRFCVersion_DRAFT_VERSION_03
+	}
+	return filter, nil
+}
+
+func modIfNoExisting(protoext proto.Message) pluginutils.ModifyFunc {
+	return func(existing *any.Any) (proto.Message, error) {
+		if existing == nil {
+			return protoext, nil
+		}
+		return nil, ErrConfigurationExists
+	}
+}
+
+func ConfigureVirtualHostFilter(settings *local_ratelimit.Settings, localRatelimit *local_ratelimit.TokenBucket, stage uint32, out *envoy_config_route_v3.VirtualHost) error {
+	filter, err := GenerateHTTPFilter(settings, localRatelimit, stage)
+	if err != nil {
+		return err
+	}
+	// Mark this stage as having user-defined configuration
+	return pluginutils.ModifyVhostPerFilterConfig(out, HTTPFilterName, modIfNoExisting(filter))
+}
+
+func (p *plugin) ProcessVirtualHost(
+	params plugins.VirtualHostParams,
+	in *v1.VirtualHost,
+	out *envoy_config_route_v3.VirtualHost,
+) error {
+	if limits := in.GetOptions().GetRatelimit().GetLocalRatelimit(); limits != nil {
+		err := ConfigureVirtualHostFilter(params.HttpListener.GetOptions().GetHttpLocalRatelimit(), limits, CustomStageBeforeAuth, out)
+		if err != nil {
+			return err
+		}
+		p.filterRequiredForListener[params.HttpListener] = struct{}{}
+		return nil
+	}
+	return nil
+}
+
+func ConfigureRouteHostFilter(settings *local_ratelimit.Settings, localRatelimit *local_ratelimit.TokenBucket, stage uint32, out *envoy_config_route_v3.Route) error {
```

Comment:
```suggestion
func ConfigureRouteFilter(settings *local_ratelimit.Settings, localRatelimit *local_ratelimit.TokenBucket, stage uint32, out *envoy_config_route_v3.Route) error {
```
can you add a comment that this func is exported for use in the enterprise plugin? (so that we don't accidentally un-export it later)

Reply:
Good catch. I've moved the helper methods to utils.go and added the comments there